fast64_internal/f3d/f3d_gbi.py

```diff
@@ -2711,6 +2752,133 @@ def save_binary(self, romfile, f3d, segments):
+
+    def bleed_mat(self, mat, LastMat):
+        if LastMat:
+            GfxList = mat.material
+            # deep copy breaks on Image objects so I will only copy the levels needed
+            commands_bled = copy.copy(GfxList)
+            commands_bled.commands = copy.copy(GfxList.commands)  # copy the commands also
+            LastList = LastMat.material.commands
+            cnt = 0
+            for j, cmd in enumerate(GfxList.commands):
+                if cmd.bleed(LastList, commands_bled.commands, j - cnt):
+                    if cmd in LastList:
+                        commands_bled.commands.pop(j - cnt)  # list gets smaller as I pop, so modify index by num popped
+                        cnt += 1
+            # remove SPEndDisplayList
+            while SPEndDisplayList() in commands_bled.commands:
+                commands_bled.commands.remove(SPEndDisplayList())
+        else:
+            commands_bled = mat.material
+        # remove SPEndDisplayList
+        while SPEndDisplayList() in commands_bled.commands:
+            commands_bled.commands.remove(SPEndDisplayList())
+        return commands_bled
+
+    def bleed_mesh_defaults(self):
+        # draw.name is used for geo layouts, but I want a separate name for the bleed revert, so I will use this new prop
+        self.draw.comment_name = f"{self.name}_revert_render_settings"
+        # remove DPPipeSync
+        while DPPipeSync() in self.draw.commands:
+            self.draw.commands.remove(DPPipeSync())
+        return self.draw
+
+    def to_c_inline(self, f3d, gfxFormatter):
+        GfxData = CData()
+        staticData = CData()
+        dynamicData = CData()
+        GfxData.header = f"extern Gfx {self.name}[];\n"
+        GfxData.source = f"Gfx {self.name}[] = {{\n"
+        size = 0
+        for triGroup in self.triangleGroups:
+            # add in vertex data
+            staticData.append(triGroup.vertexList.to_c())
+            # add mat, add tex data, then add tri data
+            if triGroup.bleed.bled_mats:
+                GfxData.append(triGroup.bleed.bled_mats.to_c_inline(f3d))
+                size += triGroup.bleed.bled_mats.size(f3d)
+            tex_size = 0
+            for tex in triGroup.bleed.bled_tex:
+                if tex:
+                    GfxData.append(tex.to_c_inline(f3d))
+                    tex_size += tex.size(f3d)
+            # if there is a scroll, change the TriGroup FMaterial to be a new one for scrolling
+            if triGroup.bleed.bleed_tile_scroll:
+                scr_mat = FMaterial("scr", DLFormat.Static)
+                for GfxList in triGroup.bleed.bled_tex:
+                    scr_mat.material.commands.extend(GfxList.commands)
+                scr_mat.material.name = f"{self.name} + {size // 8}"
+                scr_mat.name = f"{self.name} + {size // 8}"
+                scr_mat.scrollData = triGroup.fMaterial.scrollData
+                scr_mat.texturesLoaded = triGroup.fMaterial.texturesLoaded
```

Comment:
If the tile scroll code is refactored I could avoid this, but I believe that would be more messy, or would require implementing entirely new tile scroll code separate for inline materials.
I can make an attempt at it if that's the preferred direction but that was also the initial way I tried when making this PR, and it was much more complicated.

Reply:
This current solution is already pretty messy itself. The original tile scrolling code is pretty inflexible, but unless you want to refactor that too then its better to handle inline tiles scrolls as a separate case, which it is.

Reply:
I believe the best way to change it would still be to make edits to TriGroup data, though instead of making a new material I can edit the scroll generation to make use of inline attributes so that the correct data is generated.
Here is the list of items Id need to change in tile scroll generation to work with inline scrolls.
1. The function naming convention will need to be remade to come from the TriGroup/FMesh.
2. The cmds parsed to find the scroll index will have to come from TriGroup.bleed or some other TriGroup attribute.
3. The offsets used in the tile scroll will have to be relative to the start of the TriGroups parsed cmds, TriGroup.tile_offset perhaps.

Overall I think it would be similar to the current implementation except I would be moving from editing the material in the TriGroup to editing the TriGroup itself and the scroll generation

Reply:
Why not calculate the offsets in `bleed()` as you're building the display list? This is what I was referring to when talking about having a OrderedDict[FMaterial, tileCommandIndices] as an FMesh property, which would store indices into the inlined DL. This way you don't need any extra TriGroup properties at all.

Reply:
There is no reason it can't be done I just feel like it makes more sense to have it be defined in the TriGroup because that is the object that holds the material and generates the individual scroll.
It would also make the new inline code only require a few differences from the original code  which I believe will make it easier to maintain in the future.

Reply:
I'm not sure I understand - once you inline the DL, you're storing indices into `FMesh.draw`. These indices don't refer to any particular FTriGroup.

Reply:
I would be storing the indices in the bleed_gfx object, which is a member of the triGroup. As far as material bleeding goes, it is based on triGroups, and the material is unique to that triGroup.
It is correct that the new display list is a part of the FMesh object now, but the owner of the bled gfx is the triGroup, which is where the tile indices will go. To me it makes more sense to synchronizes one scroll to each material (which is what already happens) than to make entirely new behavior that places all scrolls from the given indices in a FMesh.